designs/switch-prometheus-to-opentelemetry.md

```diff
@@ -0,0 +1,172 @@
+Switch Prometheus to OpenTelemetry
+===================
+
+## Motivation
+
+Today, controller-runtime already provides metrics accessible through the
+Prometheus client, and exposed on the `/metrics` endpoint. However, this forces
+components to use Prometheus, when they may want to monitor their app another
+way, not necessarily with a pull system.
+
+## Goals
+
+* Provide a generic way to send metrics, so they can be exported virtually anywwhere.
+* Do not introduce existing metrics changes for users
+
+## Proposal
+
+We will entirely remove the current metrics server. If users need to run
+Prometheus, they will still be able to do so after configuring OpenTelemetry by
+using the included exporter.
+
+We will configure a global meter so we can send metrics from anywhere within
+the component. That meter will be set when needed, and not at boot time, so
+users can set their own `MeterProvider`  and specify where they need to exporte
+their metrics.
+
+A new `pkg/meter` package will introduce a `GetMeter()` method, which allows
+retrieving the component-wide meter.
+
+```
+func GetMeter() metric.Meter {
+  return otel.Meter("kubernetes.io/controller-runtime")
+}
+```
+
+This method will be aliased as `ctrl.GetMeter()`.
+
+Then, every metric we transmit across the component will switch from the
+prometheus client to that meter.
+
+For example, in `pkg/internal/controller/metrics`, we have the `ReconcileTotal`
+metric, which currently is defined like this:
+
+```go
+ReconcileTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+  Name: "controller_runtime_reconcile_total",
+  Help: "Total number of reconciliations per controller",
+}, []string{"controller", "result"})
+```
+
+This package will be removed, as we can't initialize metrics at boot time anymore.
+However, the `initMetrics` method in `pkg/internal/controller` will initialize the metric:
+
+```
+var (
+  reconcileTotal metric.Int64Counter
+)
+
+func (c *Controller) initMetrics() {
+  reconcileTotal = GetMeter().NewInt64Counter(
```

Comment:
This looks like a `Counter` and not a `CounterVec` to me.

I would think that there is a good chance that controller-runtime users are familiar with the prometheus client lib, as its the standard format used in the kube ecosystem, and less familiar with the `otel` libs. This change requires everyone who adds metrics themselves to get familiar with the new client libs and change their code. Can that be avoided?

Reply:
If your main concern is "users will have to change their syntax", I won't even try to convince you this change is worthwile.
However, I just want to note that OpenTelemetry is a CNCF project backed by most of the biggest telemetry providers (Google, Lightstep, Honeycomb and others), not a dark library that I'm trying to promote.